inc/class-tool.php

```diff
@@ -660,4 +671,227 @@ function ( $var ) {
+
+		if ( preg_match( '/^[0-9]*$/', $blog_name ) ) {
+			$blog_name .= 'a';
+		}
+
+		if ( in_array( $blog_name, $illegal_names, true ) ) {
+			$this->ok = false;
+			$this->message = __( 'Sorry, the activity name uses a reserved word', 'pressbooks-lti-provider' );
+			$this->handleRequest();
+			return '';
+		}
+
+		if ( strlen( $blog_name ) < $minimum_site_name_length ) {
+			$blog_name = str_pad( $blog_name, 4, '1' );
+		}
+
+		if ( is_subdomain_install() ) {
+			$host = wp_parse_url( esc_url( $domain ), PHP_URL_HOST );
+			$host = explode( '.', $host );
+			if ( count( $host ) > 2 ) {
+				array_shift( $host );
+			}
+			$bare_domain = implode( '.', $host );
+			$my_domain   = $blog_name . '.' . $bare_domain;
+			$path        = $base;
+
+		} else {
+			$illegal_sub_directory_names = array_merge( $illegal_names, get_subdirectory_reserved_names() );
+			if ( in_array( $blog_name, $illegal_sub_directory_names, true ) ) {
+				$this->ok = false;
+				$this->message = __( 'Sorry, the activity name uses a reserved word', 'pressbooks-lti-provider' );
+				$this->handleRequest();
+				return '';
+			}
+			$my_domain = "$domain";
+			$path      = $base . $blog_name . '/';
+
+		}
+
+		$path = ( 0 === strcmp( $path, '/' ) ) ? '' : $path;
+
+		return sprintf( '%1$s://%2$s%3$s', $scheme, $my_domain, $path );
+	}
+
+	/**
+	 * Creates a unique URL based on a given URL
+	 *
+	 * @param $url
+	 *
+	 * @return string
+	 * @since 1.4.0
+	 */
+	public function maybeDisambiguateDomain( $url ) {
+		$parts = wp_parse_url( $url );
+		if ( ! isset( $parts['host'] ) ) {
+			return '';
+		}
+
+		$domain = $parts['host'];
+		$path = $parts['path'];
+
+		if ( is_subdomain_install() ) {
+			$i = 1;
+			while ( domain_exists( $domain, $parts['path'], 1 ) && $i < 1000 ) {
+				$domain = "{$parts['host']}{$i}";
+				++ $i;
+			}
+		} else {
+			$i = 1;
+			while ( domain_exists( $parts['host'], $path, 1 ) && $i < 1000 ) {
+				$path = untrailingslashit( $parts['path'] ) . $i;
+				++ $i;
+			}
+		}
+
+		return sprintf( '%1$s://%2$s%3$s', $parts['scheme'], $domain, untrailingslashit( $path ) );
+
+	}
+
+	/**
+	 * Create a new book, launched via LTI
+	 *
+	 * @param string $new_book_url URL of the new book to be created
+	 * @param string $title Title of the book
+	 * @param int $user_id ID of the user creating the book
+	 * @param int $resource_link_id ID of the activity, derived from the Tool Consumer
+	 * @param int $context_id ID of the course the activity belongs to, derived from the Tool Consumer
+	 *
+	 * @return int|\WP_Error
+	 * @since 1.4.0
+	 *
+	 */
+	public function createNewBook( $new_book_url, $title, $user_id, $resource_link_id, $context_id ) {
+		$url    = untrailingslashit( $new_book_url );
+		$domain = wp_parse_url( $url, PHP_URL_HOST );
+		$path   = wp_parse_url( $url, PHP_URL_PATH );
+
+		$book_id = wpmu_create_blog( $domain, $path, $title, $user_id );
+
+		add_blog_option(
+			$book_id, 'pressbooks_lti_consumer_context', [
```

Comment:
there are several occurrences of `pressbooks_lti_consumer_context`, pls convert it to a class constant and if you need it to use it outside of the class, invoke it statically.

Reply:
great suggestion, thanks